packages/core/src/domain/telemetry/telemetry.ts

```diff
@@ -0,0 +1,236 @@
+  _: any,
+  __: string,
+  descriptor: TypedPropertyDescriptor<T>
+) {
+  const originalMethod = descriptor.value!
+  descriptor.value = function (this: any, ...args: Parameters<T>) {
+    const decorated = onRawTelemetryEventCollected ? monitor(originalMethod) : originalMethod
+    return decorated.apply(this, args) as ReturnType<T>
+  } as T
+}
+
+export function monitor<T extends (...args: any[]) => any>(fn: T): T {
+  return function (this: any) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+    return callMonitored(fn, this, arguments as unknown as Parameters<T>)
+  } as unknown as T // consider output type has input type
+}
+
+export function callMonitored<T extends (...args: any[]) => any>(
+  fn: T,
+  context: ThisParameterType<T>,
+  args: Parameters<T>
+): ReturnType<T> | undefined
+export function callMonitored<T extends (this: void) => any>(fn: T): ReturnType<T> | undefined
+export function callMonitored<T extends (...args: any[]) => any>(
+  fn: T,
+  context?: any,
+  args?: any
+): ReturnType<T> | undefined {
+  try {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+    return fn.apply(context, args)
+  } catch (e) {
+    logErrorIfDebug(e)
+    try {
+      addTelemetryError(e)
+    } catch (e) {
+      logErrorIfDebug(e)
+    }
+  }
+}
+
+export function addTelemetryDebug(message: string, context?: Context) {
+  logDebugIfDebug(message, context)
+  addTelemetry(
+    assign(
+      {
+        message,
+        status: StatusType.debug,
+      },
+      context
+    )
+  )
+}
+
+export function addTelemetryError(e: unknown) {
+  addTelemetry(
+    assign(
+      {
+        status: StatusType.error,
+      },
+      formatError(e)
+    )
+  )
+}
+
+function addTelemetry(event: RawTelemetryEvent) {
+  if (onRawTelemetryEventCollected && telemetryConfiguration.sentEventCount < telemetryConfiguration.maxEventsPerPage) {
+    telemetryConfiguration.sentEventCount += 1
+    onRawTelemetryEventCollected(event)
+  }
+}
+
+function formatError(e: unknown) {
+  if (e instanceof Error) {
+    const stackTrace = computeStackTrace(e)
+    return {
+      error: {
+        kind: stackTrace.name,
+        stack: toStackTraceString(scrubCustomerFrames(stackTrace)),
+      },
+      message: stackTrace.message!,
+    }
+  }
+  return {
+    error: {
+      stack: 'Not an instance of error',
+    },
+    message: `Uncaught ${jsonStringify(e)!}`,
+  }
+}
+
+export function scrubCustomerFrames(stackTrace: StackTrace) {
+  stackTrace.stack = stackTrace.stack.filter(
+    (frame) => !frame.url || ALLOWED_FRAME_URLS.some((allowedFrameUrl) => startsWith(frame.url!, allowedFrameUrl))
+  )
+  return stackTrace
+}
+
+export function setDebugMode(debugMode: boolean) {
+  telemetryConfiguration.debugMode = debugMode
+}
+
+function logErrorIfDebug(e: any) {
+  if (telemetryConfiguration.debugMode) {
+    display.error('[TELEMETRY ERROR]', e)
+  }
+}
+
+function logDebugIfDebug(message: any, context?: Context) {
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏This function name is a bit awkward. What about having something like:

```ts
function logIfDebug(api: ConsoleApiName, ...args: any[]) {
  if (telemetryConfiguration.debugMode) {
    display(api, '[TELEMETRY]', ...args)
  }
}
```

The log level is represented by the console API used